producer/writer/consumer_service_writer.go

```diff
@@ -0,0 +1,300 @@
+		func(d producer.Data) bool {
+			return true
+		},
+	)
+)
+
+type consumerServiceWriter interface {
+	// Write writes data.
+	Write(d producer.RefCountedData) error
+
+	// Init will initialize the consumer service writer.
+	Init()
+
+	// Close closes the writer and the background watch thread.
+	// It should block until all the data for the given consumer service
+	// have been consumed.
+	Close()
+
+	// RegisterFilter registers a filter for the consumer service.
+	RegisterFilter(fn producer.FilterFunc)
+
+	// UnregisterFilter unregisters the filter for the consumer service.
+	UnregisterFilter()
+
+	// ServiceID returns the service id of the writer.
+	ServiceID() services.ServiceID
+}
+
+type consumerServiceWriterMetrics struct {
+	placementError  tally.Counter
+	placementUpdate tally.Counter
+	invalidShard    tally.Counter
+	filteredPassed  tally.Counter
+	filterNotPassed tally.Counter
+}
+
+func newConsumerServiceWriterMetrics(m tally.Scope) consumerServiceWriterMetrics {
+	return consumerServiceWriterMetrics{
+		placementUpdate: m.Counter("placement-update"),
+		placementError:  m.Counter("placement-error"),
+		invalidShard:    m.Counter("invalid-shard"),
+		filteredPassed:  m.Counter("filter-passed"),
+		filterNotPassed: m.Counter("filter-not-passed"),
+	}
+}
+
+type consumerServiceWriterImpl struct {
+	sync.RWMutex
+
+	cs           topic.ConsumerService
+	numShards    uint32
+	ps           placement.Service
+	shardWriters []shardWriter
+	opts         Options
+	logger       log.Logger
+
+	value           runtime.Value
+	dataFilter      producer.FilterFunc
+	router          ackRouter
+	consumerWriters map[string]consumerWriter
+	closed          bool
+	m               consumerServiceWriterMetrics
+
+	processFn runtime.ProcessFn
+}
+
+// TODO: Remove the nolint comment after adding usage of this function.
+// nolint: deadcode
+func newConsumerServiceWriter(
+	cs topic.ConsumerService,
+	numShards uint32,
+	mPool messagePool,
+	opts Options,
+) (consumerServiceWriter, error) {
+	ps, err := opts.ServiceDiscovery().PlacementService(cs.ServiceID(), nil)
+	if err != nil {
+		return nil, err
+	}
+	router := newAckRouter(int(numShards))
+	w := &consumerServiceWriterImpl{
+		cs:              cs,
+		numShards:       numShards,
+		ps:              ps,
+		shardWriters:    initShardWriters(router, cs, numShards, mPool, opts),
+		opts:            opts,
+		logger:          opts.InstrumentOptions().Logger(),
+		dataFilter:      acceptAllFilter,
+		router:          router,
+		consumerWriters: make(map[string]consumerWriter),
+		closed:          false,
+		m:               newConsumerServiceWriterMetrics(opts.InstrumentOptions().MetricsScope()),
+	}
+	w.processFn = w.process
+	return w, nil
+}
+
+func initShardWriters(
+	router ackRouter,
+	cs topic.ConsumerService,
+	numberOfShards uint32,
+	mPool messagePool,
+	opts Options,
+) []shardWriter {
+	sws := make([]shardWriter, numberOfShards)
+	for i := range sws {
+		switch cs.ConsumptionType() {
```

Comment:
Saw that there's an Unknown Consumption type. Will there be issues if there's no default case for this? Also seems like since cs.ConsumptionType() is a single value, perhaps there's no need for the switch statement for every iteration in sws. Perhaps can declare the consumption type above and do a if statement.

Reply:
sounds good, will handle the unknown case